src/util/fields.py

```diff
@@ -631,6 +632,71 @@ def convert_field_to_html(cr, model, field, skip_inherit=()):
         convert_field_to_html(cr, inh.model, field, skip_inherit=skip_inherit)
 
 
+def convert_field_to_company_dependent(
+    cr, model, field, type, target_model=None, default_value=None, default_value_ref=None, company_field="company_id"
+):
+    _validate_model(model)
+    if target_model:
+        _validate_model(target_model)
+
+    type2field = {"char", "float", "boolean", "integer", "text", "many2one", "date", "selection", "html"}
+    assert type in type2field
+
+    table = table_of_model(cr, model)
+
+    cr.execute("SELECT id FROM ir_model_fields WHERE model=%s AND name=%s", (model, field))
+    if not cr.rowcount:
+        # TODO: cwg not sure what to do
+        return
+    [field_id] = cr.fetchone()
+
+    params = []
+    _field = f"_{field}"
+    if default_value is None:
+        where_condition = f"{_field} IS NOT NULL"
+    else:
+        where_condition = f"{_field} != %s"
+        params.append(default_value)
+
+    cr.execute(f"ALTER TABLE {table} RENAME COLUMN {field} TO {_field}")
+
+    if is_field_anonymized(cr, model, field):
+        # TODO: cwg not sure what to do
+        pass
+
+    cr.execute(f"ALTER TABLE {table} ADD COLUMN {field} jsonb")
+    cr.execute(
+        f"""
+        UPDATE {table}
+           SET {field} = jsonb_build_object({company_field}, {_field})
+         WHERE {where_condition}
+        """,
+        params,
+    )
+    cr.execute(f"ALTER TABLE {table} DROP COLUMN {_field}")
```

Comment:
Use `util.alter_column_type`

Reply:
I cannot directly use `alter_column_type`
1. the logic ```remove the existing linked `ir_model_fields_selection` recods in case it was a selection field``` shouldn't be executed, since changing a selection field to company dependent field shouldn't remove all its selection options
2. I need a parameter `where` to avoid migrate data which is the same as the default value

I think the current `alter_column_type` is a good and simple logic abstraction, it would be better to not modify it and just copy the similar code to `convert_field_to_company_dependent`

What do you think?

Reply:
1. This is not a problem. The `ir_model_fields_selection` records are not required. They will be recreated by the ORM when reflecting the field. You can also modify this function to skip this step conditionally.
2. it could be included in the `using` clause by wrapping the `jsonb_build_object` around a `CASE WHEN`.

Reply:
Another problem is the SQL requires variable for `default`
I guess `explode_execute` doesn't support params?

Reply:
You just call `cr.mogrify` to inject params.